CMakeLists.txt

```diff
@@ -1,19 +1,57 @@
-cmake_minimum_required(VERSION 3.1)
+cmake_minimum_required(VERSION 3.8)
 
-project(dbg_macro)
+project(dbg_macro LANGUAGES CXX)
 
-set(CMAKE_CXX_STANDARD 11)
+###############################################################################
+# CMake Options
+###############################################################################
+
+option(DBG_MACRO_BUILD_EXAMPLES OFF)
+option(DBG_MACRO_BUILD_TESTS OFF)
+
+###############################################################################
+# Add dbg_macro Library
+###############################################################################
 
 add_library(${PROJECT_NAME} INTERFACE)
-target_include_directories(${PROJECT_NAME} INTERFACE ".")
+add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
+target_include_directories(${PROJECT_NAME} INTERFACE
+	$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+	$<INSTALL_INTERFACE:include>)
+target_compile_features(${PROJECT_NAME} INTERFACE cxx_std_11)
+
+###############################################################################
+# Build Examples
+###############################################################################
+
+if(DBG_MACRO_BUILD_EXAMPLES)
+	macro(dbg_macro_add_example EXAMPLENAME)
+		add_executable(${EXAMPLENAME} ${ARGN})
+		target_link_libraries(${EXAMPLENAME} PRIVATE dbg_macro::dbg_macro)
+		set_target_properties(${EXAMPLENAME} PROPERTIES FOLDER examples)
+	endmacro()
+	add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/examples")
+endif()
+
+###############################################################################
+# Build Tests
+###############################################################################
 
-add_executable(${PROJECT_NAME}-example tests/example.cpp)
-target_compile_options(${PROJECT_NAME}-example PRIVATE -Wall -Wextra -pedantic)
-target_link_libraries(${PROJECT_NAME}-example ${PROJECT_NAME})
+if(DBG_MACRO_BUILD_TESTS)
+	enable_testing()
+	macro(dbg_macro_add_test TESTNAME)
+		add_executable(${TESTNAME} ${ARGN})
+		target_link_libraries(${TESTNAME} PRIVATE dbg_macro::dbg_macro)
+		add_test(NAME ${TESTNAME} COMMAND ${TESTNAME})
+		set_target_properties(${TESTNAME} PROPERTIES FOLDER tests)
+	endmacro()
+	add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/tests")
+endif()
 
-add_executable(${PROJECT_NAME}-tests tests/tests.cpp)
-target_compile_options(${PROJECT_NAME}-tests PRIVATE -Wall -Wextra -pedantic)
-target_link_libraries(${PROJECT_NAME}-tests ${PROJECT_NAME})
+###############################################################################
+# Install
```

Comment:
what exactly does this block achieve? What is `dbg_macroConfig`?

Reply:
I am not really sure (@Super Wig recommended this) but this probably allows users to use the `find_package(dbg_macro)` command to be used with cmake projects.

Reply:
It enables support for
```cmake
find_package(dbg_macro)
```
CMake will look for a dbg_macro-config.cmake or dbg_macroConfig.cmake file which is used to add everything needed in order to use the target. With this library that's not a whole lot as it's just cxx_std_11 and the include directory, it just makes it easier to integrate as all a user needs to do is
```cmake
find_package(dbg_macro)
target_link_libraries(target PRIVATE dbg_macro::dbg_macro)
```
----
So for a user to install it's
```
git clone https://github.com/sharkdp/dbg-macro
mkdir dbg-macro/build && cd dbg-macro/build
cmake ..
sudo make install //or "cmake --build . --target install" if you want to be agnostic
```
And then this will put the header in `/usr/local/include/dbg_macro` and the config in `/usr/local/lib/cmake`

Reply:
Thank you for the detailed explanation!